### PR TITLE
Fixes support for TcpRoute in the AppNet APIs

### PR DESCRIPTION
**Pull request by bowei**
Fix various bugs needed to make TcpRoute resource work
    
    - Operations needs to be implemented
    - Various mismatches in the interface with the compute API

### CODE REVIEW

File: e2e/tcproute_test.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	"fmt"
+	"testing"
+
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/meta"
+	"github.com/kr/pretty"
+	"google.golang.org/api/compute/v1"
+	"google.golang.org/api/networkservices/v1"
+)
+
+func TestTcpRoute(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (bowei) [reply e2e/tcproute_test.go:29]: These are run manually for now. We have a background investigation to setup CI, but it was a bit complex.

[the remark this answers is not in the document]
Comment (bowei) [reply e2e/tcproute_test.go:29]: (e.g. it is work in progress)

**Comment by bowei** [comment]
This should be ready for review

File: e2e/tcproute_test.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+	// Insert
+	err = theCloud.TcpRoutes().Insert(ctx, key, tcpr)
+	t.Logf("tcproutes insert: %v", err)
+	if err != nil {
+		t.Fatalf("Insert() = %v", err)
+	}
+
+	// Get
+	tcpRoute, err := theCloud.TcpRoutes().Get(ctx, key)
```
Comment (aojea) [inline e2e/tcproute_test.go:91]: don't you need to assert that what you get is really what you inserted?

File: pkg/cloud/meta/service.go [hunk 3]
```diff
@@ -116,6 +117,19 @@ func (i *ServiceInfo) ListItemName() string {
 	return "Items"
 }
 
+func (i *ServiceInfo) NetworkServicesFmt() string {
```
Comment (aojea) [inline pkg/cloud/meta/service.go:120]: Why don't you use the Stringer https://go.dev/tour/methods/17 implementing the `String()`method for this?

File: pkg/cloud/op_networkservices.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+type networkServicesOperation struct {
+	s         *Service
+	projectID string
+	key       *meta.Key
+	err       error
+}
+
+func (o *networkServicesOperation) String() string {
+	return fmt.Sprintf("networkServicesOperation{%q, %v}", o.projectID, o.key)
```
Comment (aojea) [inline pkg/cloud/op_networkservices.go:39]: `o.key` or `o.key.String()`

File: pkg/cloud/op_networkservices.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	return &RateLimitKey{
+		ProjectID: o.projectID,
+		Operation: "Get",
+		Service:   "Operations",
+		Version:   meta.VersionGA,
+	}
+}
+
+func (o *networkServicesOperation) error() error {
+	return o.err
```
Comment (aojea) [inline pkg/cloud/op_networkservices.go:83]: do we care about thread safety here?

File: pkg/cloud/service.go [hunk 6]
```diff
@@ -41,32 +47,122 @@ type Service struct {
...
 
+type networkServiceOpURLParseResult struct {
+	projectID string
+	key       *meta.Key
+}
+
+// parseNetworkServiceOpURL parses the URL of the network services operation.
+// This is different than the `compute` API paths.
+func parseNetworkServiceOpURL(name string) (*networkServiceOpURLParseResult, error) {
+	// Format: projects/<projectID>/locations/global/operations/<Name>
```
Comment (aojea) [inline pkg/cloud/service.go:154]: do we need to do a sanity check that split[0] == projects, split[2] == locations , .... ?

File: e2e/tcproute_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (aojea) [reply e2e/tcproute_test.go:29]: these tests requires to perform GCE calls or are mocked?

**Comment by aojea** [comment]
mostly questions for self education

File: pkg/cloud/service.go (hunk 6 above)
Comment (bowei) [reply pkg/cloud/service.go:154]: adding.

File: pkg/cloud/op_networkservices.go (hunk 5 above)
Comment (bowei) [reply pkg/cloud/op_networkservices.go:83]: library is mostly thread agnostic -- you are expected to guard for threads outside this code.

File: pkg/cloud/op_networkservices.go (hunk 4 above)
Comment (bowei) [reply pkg/cloud/op_networkservices.go:39]: changed to use %s, which calls String()

File: pkg/cloud/meta/service.go (hunk 3 above)
Comment (bowei) [reply pkg/cloud/meta/service.go:120]: Don't completely follow your comment, can you explain more?

**Pull request merged**